Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbInputDeviceKeyboard.cpp

```diff
@@ -268,4 +327,33 @@ namespace AzFramework
             default: return nullptr;
         }
     }
+
+    AZStd::string XcbInputDeviceKeyboard::TextFromKeycode(xkb_state *state, xkb_keycode_t code)
+    {
+        // Find out how much of a buffer we need
+        const size_t size = xkb_state_key_get_utf8(state, code, nullptr, 0);
+        if (!size)
+        {
+            return {};
+        }
+        // xkb_state_key_get_utf8 will null-terminate the resulting string, and
+        // will truncate the result if there is not enough space for the null
+        // byte. The first call returns the size of the resulting string
+        // without including the null byte. Reserve enough space for the string
+        // + null byte, set the string's size, then use xkb_state_key_get_utf8
+        // to copy the string data
+        AZStd::string chars;
+        chars.reserve(size + 1);
+        chars.resize(size);
+        xkb_state_key_get_utf8(state, code, chars.data(), chars.size() + 1);
```

Comment:
Yes, all that is maintained in that `state` parameter:
> State objects contain the active state of a keyboard (or keyboards), such as the currently effective layout and the active modifiers.

Part of this change is to update the keyboard support to react to `XCB_XKB_STATE_NOTIFY` events properly, so that our local state is updated when modifier keys are pressed and released. The only part that is still missing is code to react to `XCB_XKB_MAP_NOTIFY` events, to update the keymap when the user changes the active map (with `setxkbmap` or something else) _while the game is running_. It does already properly initialize the state using the active keymap on launch.

Reply:
Great!